core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -389,12 +389,6 @@ public long cleanUpAllAllocatedMemory() {
       consumers.clear();
     }
 
-    for (MemoryBlock page : pageTable) {
```

Comment:
Chatted with @JoshRosen , these lines are introduce by logical conflict, it could be rewritten as
```scala
    synchronized (this) {
      for (MemoryConsumer c: consumers) {
        if (c != null && c.getUsed() > 0) {
          // In case of failed task, it's normal to see leaked memory
          logger.warn("leak " + Utils.bytesToString(c.getUsed()) + " memory from " + c);
        }
      }
      consumers.clear();

      for (MemoryBlock page : pageTable) {
        if (page != null) {
          logger.warn("leak a page: " + page + " in task " + taskAttemptId);
          memoryManager.tungstenMemoryAllocator().free(page);
        }
      }
      Arrays.fill(pageTable, null);
    }
```